src/ucp/rma/basic_rma.c

```diff
@@ -26,142 +30,242 @@
+}
 
-        req->send.buffer += packed_len;
-        req->send.rma.remote_addr += packed_len;
-        return UCS_INPROGRESS;
-    } else {
+static ucs_status_t 
+ucp_progress_get_inner(ucp_request_t *req, uct_rkey_t uct_rkey, 
+                       ucp_ep_rma_config_t *rma_config)
+{
+    ucp_ep_t *ep = req->send.ep;
+    ucs_status_t status;
+    size_t frag_length;
+
+    ++req->send.uct_comp.count;
+    frag_length = ucs_min(rma_config->max_get_bcopy, req->send.length);
+    status = UCS_PROFILE_CALL(uct_ep_get_bcopy,
+                              ep->uct_eps[req->send.lane],
+                              (uct_unpack_callback_t)memcpy,
+                              (void*)req->send.buffer,
+                              frag_length,
+                              req->send.rma.remote_addr,
+                              uct_rkey,
+                              &req->send.uct_comp);
+    if (status <= 0) {
+        --req->send.uct_comp.count;
+    }
+
+    return ucp_rma_request_advance(req, frag_length, status);    
+}
+
+static ucs_status_t ucp_progress_get_nbi(uct_pending_req_t *self)
+{
+    uct_rkey_t uct_rkey;
+    ucp_ep_rma_config_t *rma_config;
+    ucs_status_t status;
+
+    ucp_request_t *req = ucs_container_of(self, ucp_request_t, send.uct);
+    ucp_rkey_h rkey    = req->send.rma.rkey;
+    ucp_ep_t *ep       = req->send.ep;
+
+    UCP_EP_RESOLVE_RKEY_RMA(ep, rkey, req->send.lane, uct_rkey, rma_config);
+    status = ucp_progress_get_inner(req, uct_rkey, rma_config);
+
+    return status;
+}
+
+static ucs_status_t ucp_progress_put_nbi(uct_pending_req_t *self)
+{
+    uct_rkey_t uct_rkey;
+    ucp_ep_rma_config_t *rma_config;
+    ucs_status_t status;
+
+    ucp_request_t *req = ucs_container_of(self, ucp_request_t, send.uct);
+    ucp_rkey_h rkey    = req->send.rma.rkey;
+    ucp_ep_t *ep       = req->send.ep;
+
+    UCP_EP_RESOLVE_RKEY_RMA(ep, rkey, req->send.lane, uct_rkey, rma_config);
+    status = ucp_progress_put_inner(req, uct_rkey, rma_config);
+
+    return status;
+}
+
+static inline ucs_status_t
+ucp_rma_blocking(ucp_ep_h ep, const void *buffer, size_t length, uint64_t remote_addr, 
+                 ucp_rkey_h rkey, ucp_ep_rma_config_t *rma_config, 
+                 ucp_lane_index_t lane, uct_rkey_t uct_rkey,
+                 ucp_rma_send_func_t send_func)
+{
+    ucs_status_t status;
+    ucp_request_t req;
+
+    ucs_assert((send_func == ucp_progress_put_inner) ||
+               (send_func == ucp_progress_get_inner));
+
+    status = ucp_rma_request_init(&req, ep, buffer, length, remote_addr, rkey,
+                                  NULL, lane, 0);
+    if (status != UCS_OK) {
         return status;
     }
+
+    /* Loop until all message has been sent.
+     * We re-check the configuration on every iteration except for zcopy, 
+     * because it can be * changed by transport switch.
+     */
+    for (;;) {
+        status = send_func(&req, uct_rkey, rma_config);
```

Comment:
pls check in disassembly if the function pointer is optimized (for example, using "perf top" or "objdump -ld libucp.so")